File: pset6/credit.py
from cs50 import get_int

# Main function


def main():
    # Ask for input of card number
    card_num = get_int("Number: ")
    num_of_digits = get_card(card_num)
    sum = check_sum(card_num, num_of_digits)
    first_two = get_digit(card_num, num_of_digits) * 10 + get_digit(card_num, num_of_digits - 1)
    if sum % 10 == 0:
        if (first_two == 34 or first_two == 37) and num_of_digits == 15:
            print("AMEX")
            exit(0)
        elif (first_two > 50 and first_two < 56) and num_of_digits == 16:
            print("MASTERCARD")
            exit(0)
        elif int(first_two / 10) == 4 and (num_of_digits == 13 or num_of_digits == 16):
            print("VISA")
            exit(0)
        else:
            print("INVALID")
            exit(2)
    else:
        print("INVALID")
        exit(3)


# Return the num of digits in the card number
def get_card(card_num):
    if card_num >= 1000000000000 and card_num < 10000000000000:
        return 13
    elif card_num >= 100000000000000 and card_num < 1000000000000000:
        return 15
    elif card_num >= 1000000000000000 and card_num < 10000000000000000:
        return 16
    else:
        print("INVALID")
        exit(1)


# check sum function
def check_sum(card, num_of_digits):
    sum = 0
    counter = 1
    while True:
        current = get_digit(card, counter)
        if counter % 2 == 0:
            current *= 2
            if current > 9:
                current = get_digit(current, 1) + get_digit(current, 2)
        sum += current
        counter += 1
        if counter >= (num_of_digits + 1):
            break
    return sum


# function to get a digit from a number
def get_digit(num, n):
    if (n == 1):
        return num % 10
    return int(num / 10 ** (n - 1) % 10)


# Call main function
main()

File: pset1/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

//Declaring functions
int checksum(long card);
int getdigit(long num, int n);

int main(void)
{
    //get a long int
    long cardNum = get_long("Number: ");
    int numOfDigits;
    //check if right number of digits and save the number for later use
    if (cardNum>=1000000000000 && cardNum<10000000000000)
    {
        numOfDigits=13;
    }
    else if (cardNum>=100000000000000 && cardNum<1000000000000000)
    {
        numOfDigits=15;
    }
    else if (cardNum>=1000000000000000 && cardNum<10000000000000000)
    {
        numOfDigits=16;
    }
    //if not right number of digits, card is invalid
    else
    {
        printf("INVALID\n");
        return 0;
    }
    //use check sum to find out if valid
    int sum=checksum(cardNum);
    //find the first two digits to identify which credit card structure
    int firsttwo=getdigit(cardNum, numOfDigits)*10+getdigit(cardNum, numOfDigits-1);
    //check for credit card company and correct num of digits
    if (sum%10==0) {
        if ((firsttwo==34 || firsttwo==37) && numOfDigits==15)
        {
            printf("AMEX\n");
            return 0;
        }
        else if ((firsttwo>50 && firsttwo<56) && numOfDigits==16)
        {
            printf("MASTERCARD\n");
            return 0;
        }
        else if (firsttwo/10==4 && (numOfDigits==13 || numOfDigits==16))
        {
            printf("VISA\n");
            return 0;
        }
        //if structure does not match, it is invalid
        else
        {
            printf("INVALID\n");
            return 0;
        }
    }
    else {
        printf("INVALID\n");
        return 0;
    }
}
//check sum function:
int checksum(long card)
{
    int sum=0;
    int current;
    int numOfDigits;
    if (card>=1000000000000 && card<10000000000000)
    {
        numOfDigits=13;
    }
    else if (card>=100000000000000 && card<1000000000000000)
    {
        numOfDigits=15;
    }
    else {
        numOfDigits=16;
    }
    int counter=1;
    do
    {
        current=getdigit(card, counter);
        if (counter%2==0)
        {
            current*=2;
            if (current>9)
            {
                current=getdigit(current, 1)+getdigit(current,2);
            }

        }
        sum+=current;
        counter++;
    }
    while (counter<numOfDigits+1);
    return sum;
}
//function to get a digit from a number
int getdigit(long num, int n)
{
    long answer;
    if (n==1)
    {
        return num%10;
    }
    answer=(long)(num/pow(10,(n-1)))%10;
    return answer;
}
